Sources/SwiftGRPCNIO/ClientCalls/ClientCall.swift

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright 2019, gRPC Authors All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import Foundation
+import NIO
+import NIOHTTP1
+import NIOHTTP2
+import SwiftProtobuf
+
+public protocol ClientCall {
+  associatedtype RequestMessage: Message
+  associatedtype ResponseMessage: Message
+
+  /// HTTP/2 stream that requests and responses are sent and received on.
+  var subchannel: EventLoopFuture<Channel> { get }
+
+  /// Initial response metadata.
+  var initialMetadata: EventLoopFuture<HTTPHeaders> { get }
+
+  /// Response status.
+  var status: EventLoopFuture<GRPCStatus> { get }
+
+  /// Trailing response metadata.
+  ///
+  /// This is the same metadata as `GRPCStatus.trailingMetadata` returned by `status`.
+  var trailingMetadata: EventLoopFuture<HTTPHeaders> { get }
+
+  /// Cancels the current call.
+  func cancel()
+}
+
+extension ClientCall {
+  public var trailingMetadata: EventLoopFuture<HTTPHeaders> {
+    return status.map { $0.trailingMetadata }
+  }
+}
+
+/// A `ClientCall` with server-streaming; i.e. server-streaming and bidirectional-streaming.
+public protocol StreamingRequestClientCall: ClientCall {
+  /// Sends a request to the service. Callers must terminate the stream of messages
+  /// with an `.end` event.
+  ///
+  /// - Parameter event: event to send.
+  func send(_ event: StreamEvent<RequestMessage>)
```

Comment:
The promise would only get fulfilled once the message reaches the network -- do we really want that? (It would also require a flush after writing each message if the `end` was added to the same queue.)

Reply:
Hmm, good point. Fulfilling the promise only when we send the bytes to the network helps with regards to backpressure, though. If we e.g. want to stream 10k messages of 10kB each, we don't want to dump them all at once onto the NIO layer (and lower).

Reply:
In general, the pattern of providing a promise or returning a future here is sensible. Users that don't want to wait don't have to: there's no rule that says that the promise must complete before the next send call is made.

Reply:
Thanks for elaborating, @Lukasa! One more clarification question: if I send (but do not explicitly flush) message A, then wait for that send's promise to be fulfilled before sending message B, will message B ever be sent, even though I don't explicitly flush?

Reply:
That's correct, message B will never be sent.

Providing a promise / returning a future is fine here though (providing we flush each time). For most use cases where we have multiple messages to send and we don't want to wait for each promise to succeed, calling `writeAndFlush` each time shouldn't be an issue. 

I spoke to @weissi offline, and he suggested a write-many for cases where we don't want to flush after each message.

Reply:
Unless there are strong reasons against this, I’d prefer for our code to chain these send futures by default. If users prefer not to, they can still choose to ignore the futures.